package/src/utils/add-dts.ts

```diff
@@ -0,0 +1,136 @@
+import { readFile, writeFile } from "node:fs/promises";
+import { relative } from "node:path";
+import { fileURLToPath } from "node:url";
+import type { AstroIntegrationLogger } from "astro";
+import { useHookParams } from "../internal.js";
+import { ensureDirExists } from "../internal/node.js";
+
+const injectEnvDTS = async ({
+	srcDir,
+	logger,
+	specifier,
+}: {
+	srcDir: URL;
+	logger: AstroIntegrationLogger;
+	specifier: URL | string;
+}) => {
+	const envDTsPath = fileURLToPath(new URL("env.d.ts", srcDir));
+
+	if (specifier instanceof URL) {
+		specifier = fileURLToPath(specifier);
+		specifier = relative(fileURLToPath(srcDir), specifier);
+		specifier = specifier.replaceAll("\\", "/");
+	}
+
+	const envDTsContents = await readFile(envDTsPath, "utf8");
+
+	if (envDTsContents.includes(`/// <reference types='${specifier}' />`)) {
+		return;
+	}
+	if (envDTsContents.includes(`/// <reference types="${specifier}" />`)) {
+		return;
+	}
+
+	const newEnvDTsContents = envDTsContents
+		.replace(
+			`/// <reference types='astro/client' />`,
+			`/// <reference types='astro/client' />\n/// <reference types='${specifier}' />\n`,
```

Comment:
This replaces a text without a newline at the end with a text _with_ a newline at the end. So, each integration that uses this will create an extra blank line at the end of the block.